Zero initialize all pointer class and struct members

akien-mga opened the pull request:
This prevents the pitfall of UB when checking if they have been
assigned something valid by comparing to nullptr.

Alternative options for zero initialization: https://en.cppreference.com/w/cpp/language/zero_initialization
```cpp
Object *ptr = nullptr;
Object *ptr{};
Object *ptr = {};
```

When we discussed the `Object *ptr{};` syntax a couple of years ago at a GodotCon, several of us were not so comfortable with this style / found it harder to read that using the `=` operator. If that's still the case, the `Object *ptr = {};` syntax could be an alternative. But that's likely for a separate ~bikeshedding~ discussion, we can always change it later with regexps.

Also worth noting, many of the pending PRs might not be using zero init for pointers right now. We should try to point it out when reviewing, but if some changes go through with an explicit `nullptr` init, it's easy enough to re-run the regexp below.

----

Methodology for the changes:
```
find -name "thirdparty" -prune -o -name "*.h" -exec sed -i {} -e '/return /! s/\t\([A-Za-z0-9_]* \*[A-Za-z0-9_]*\)\;/\t\1 = nullptr;/g' \;
```

Then I undid the changes which didn't compile (e.g. initializing multiple members in a union, and some tricky stuff with templates). Here's the list of changes which are **not** included in this PR:

```diff
diff --git a/core/extension/gdnative_interface.h b/core/extension/gdnative_interface.h
index 732a1b5af0..746e77dde9 100644
--- a/core/extension/gdnative_interface.h
+++ b/core/extension/gdnative_interface.h
@@ -210,7 +210,7 @@ typedef struct {
 	GDNativePropertyInfo return_value;
 	uint32_t flags; // From GDNativeExtensionClassMethodFlags
 	int32_t id;
-	GDNativePropertyInfo *arguments;
+	GDNativePropertyInfo *arguments = nullptr;
 	uint32_t argument_count;
 	GDNativeVariantPtr default_arguments;
 	uint32_t default_argument_count;
@@ -241,7 +241,7 @@ typedef struct {
 	GDNativeExtensionClassFreeInstance free_instance_func; /* this one is mandatory */
 	GDNativeExtensionClassGetVirtual get_virtual_func;
 	GDNativeExtensionClassGetRID get_rid_func;
-	void *class_userdata;
+	void *class_userdata = nullptr;
 } GDNativeExtensionClassCreationInfo;
 
 typedef void *GDNativeExtensionClassLibraryPtr;
@@ -287,7 +287,7 @@ typedef GDNativeExtensionClassMethodArgumentMetadata (*GDNativeExtensionClassMet
 
 typedef struct {
 	const char *name;
-	void *method_userdata;
+	void *method_userdata = nullptr;
 	GDNativeExtensionClassMethodCall call_func;
 	GDNativeExtensionClassMethodPtrCall ptrcall_func;
 	uint32_t method_flags; /* GDNativeExtensionClassMethodFlags */
@@ -297,7 +297,7 @@ typedef struct {
 	GDNativeExtensionClassMethodGetArgumentInfo get_argument_info_func; /* name and hint information for the argument can be omitted in release builds. Class name should always be present if it applies. */
 	GDNativeExtensionClassMethodGetArgumentMetadata get_argument_metadata_func;
 	uint32_t default_argument_count;
-	GDNativeVariantPtr *default_arguments;
+	GDNativeVariantPtr *default_arguments = nullptr;
 } GDNativeExtensionClassMethodInfo;
 
 /* SCRIPT INSTANCE EXTENSION */
@@ -562,7 +562,7 @@ typedef struct {
 	 * If Core or Servers, the extension needs editor or game restart to take effect */
 	GDNativeInitializationLevel minimum_initialization_level;
 	/* Up to the user to supply when initializing */
-	void *userdata;
+	void *userdata = nullptr;
 	/* This function will be called multiple times for each initialization level. */
 	void (*initialize)(void *userdata, GDNativeInitializationLevel p_level);
 	void (*deinitialize)(void *userdata, GDNativeInitializationLevel p_level);
diff --git a/core/object/callable_method_pointer.h b/core/object/callable_method_pointer.h
index 577d4b9fbd..2cea3a34f8 100644
--- a/core/object/callable_method_pointer.h
+++ b/core/object/callable_method_pointer.h
@@ -80,7 +80,7 @@ public:
 template <class T, class... P>
 class CallableCustomMethodPointer : public CallableCustomMethodPointerBase {
 	struct Data {
-		T *instance;
+		T *instance = nullptr;
 #ifdef DEBUG_ENABLED
 		uint64_t object_id;
 #endif
@@ -134,7 +134,7 @@ Callable create_custom_callable_function_pointer(T *p_instance,
 template <class T, class R, class... P>
 class CallableCustomMethodPointerRet : public CallableCustomMethodPointerBase {
 	struct Data {
-		T *instance;
+		T *instance = nullptr;
 #ifdef DEBUG_ENABLED
 		uint64_t object_id;
 #endif
@@ -189,7 +189,7 @@ Callable create_custom_callable_function_pointer(T *p_instance,
 template <class T, class R, class... P>
 class CallableCustomMethodPointerRetC : public CallableCustomMethodPointerBase {
 	struct Data {
-		T *instance;
+		T *instance = nullptr;
 #ifdef DEBUG_ENABLED
 		uint64_t object_id;
 #endif
diff --git a/core/os/threaded_array_processor.h b/core/os/threaded_array_processor.h
index 935fc7a360..170f47f278 100644
--- a/core/os/threaded_array_processor.h
+++ b/core/os/threaded_array_processor.h
@@ -40,7 +40,7 @@ template <class C, class U>
 struct ThreadArrayProcessData {
 	uint32_t elements;
 	SafeNumeric<uint32_t> index;
-	C *instance;
+	C *instance = nullptr;
 	U userdata;
 	void (C::*method)(uint32_t, U);
 
diff --git a/core/templates/command_queue_mt.h b/core/templates/command_queue_mt.h
index 0aa1cfd541..e078bee31e 100644
--- a/core/templates/command_queue_mt.h
+++ b/core/templates/command_queue_mt.h
@@ -212,7 +212,7 @@
 #define DECL_CMD(N)                                                    \
 	template <class T, class M COMMA(N) COMMA_SEP_LIST(TYPE_PARAM, N)> \
 	struct Command##N : public CommandBase {                           \
-		T *instance;                                                   \
+		T *instance = nullptr;                                                   \
 		M method;                                                      \
 		SEMIC_SEP_LIST(PARAM_DECL, N);                                 \
 		virtual void call() {                                          \
@@ -223,8 +223,8 @@
 #define DECL_CMD_RET(N)                                                         \
 	template <class T, class M, COMMA_SEP_LIST(TYPE_PARAM, N) COMMA(N) class R> \
 	struct CommandRet##N : public SyncCommand {                                 \
-		R *ret;                                                                 \
-		T *instance;                                                            \
+		R *ret = nullptr;                                                                 \
+		T *instance = nullptr;                                                            \
 		M method;                                                               \
 		SEMIC_SEP_LIST(PARAM_DECL, N);                                          \
 		virtual void call() {                                                   \
@@ -235,7 +235,7 @@
 #define DECL_CMD_SYNC(N)                                               \
 	template <class T, class M COMMA(N) COMMA_SEP_LIST(TYPE_PARAM, N)> \
 	struct CommandSync##N : public SyncCommand {                       \
-		T *instance;                                                   \
+		T *instance = nullptr;                                                   \
 		M method;                                                      \
 		SEMIC_SEP_LIST(PARAM_DECL, N);                                 \
 		virtual void call() {                                          \
diff --git a/core/variant/callable.h b/core/variant/callable.h
index 6a760958d6..8c57247269 100644
--- a/core/variant/callable.h
+++ b/core/variant/callable.h
@@ -49,7 +49,7 @@ class Callable {
 	alignas(8) StringName method;
 	union {
 		uint64_t object = 0;
-		CallableCustom *custom;
+		CallableCustom *custom = nullptr;
 	};
 
 public:
diff --git a/core/variant/variant.h b/core/variant/variant.h
index ca18249f36..4557436185 100644
--- a/core/variant/variant.h
+++ b/core/variant/variant.h
@@ -206,12 +206,12 @@ private:
 		bool _bool;
 		int64_t _int;
 		double _float;
-		Transform2D *_transform2d;
+		Transform2D *_transform2d = nullptr;
 		::AABB *_aabb;
-		Basis *_basis;
-		Transform3D *_transform3d;
-		PackedArrayRefBase *packed_array;
-		void *_ptr; //generic pointer
+		Basis *_basis = nullptr;
+		Transform3D *_transform3d = nullptr;
+		PackedArrayRefBase *packed_array = nullptr;
+		void *_ptr = nullptr; //generic pointer
 		uint8_t _mem[sizeof(ObjData) > (sizeof(real_t) * 4) ? sizeof(ObjData) : (sizeof(real_t) * 4)]{ 0 };
 	} _data alignas(8);
 
diff --git a/modules/gdscript/gdscript_parser.h b/modules/gdscript/gdscript_parser.h
index 0da1896936..eb1c5bd347 100644
--- a/modules/gdscript/gdscript_parser.h
+++ b/modules/gdscript/gdscript_parser.h
@@ -504,11 +504,11 @@ public:
 
 			union {
 				ClassNode *m_class = nullptr;
-				ConstantNode *constant;
-				FunctionNode *function;
-				SignalNode *signal;
-				VariableNode *variable;
-				EnumNode *m_enum;
+				ConstantNode *constant = nullptr;
+				FunctionNode *function = nullptr;
+				SignalNode *signal = nullptr;
+				VariableNode *variable = nullptr;
+				EnumNode *m_enum = nullptr;
 			};
 			EnumNode::Value enum_value;
 
@@ -772,9 +772,9 @@ public:
 
 		union {
 			ParameterNode *parameter_source = nullptr;
-			ConstantNode *constant_source;
-			VariableNode *variable_source;
-			IdentifierNode *bind_source;
+			ConstantNode *constant_source = nullptr;
+			VariableNode *variable_source = nullptr;
+			IdentifierNode *bind_source = nullptr;
 		};
 		FunctionNode *source_function = nullptr;
 
@@ -869,8 +869,8 @@ public:
 
 		union {
 			LiteralNode *literal = nullptr;
-			IdentifierNode *bind;
-			ExpressionNode *expression;
+			IdentifierNode *bind = nullptr;
+			ExpressionNode *expression = nullptr;
 		};
 		Vector<PatternNode *> array;
 		bool rest_used = false; // For array/dict patterns.
@@ -933,7 +933,7 @@ public:
 		ExpressionNode *base = nullptr;
 		union {
 			ExpressionNode *index = nullptr;
-			IdentifierNode *attribute;
+			IdentifierNode *attribute = nullptr;
 		};
 
 		bool is_attribute = false;
@@ -958,9 +958,9 @@ public:
 			Type type = UNDEFINED;
 			union {
 				ConstantNode *constant = nullptr;
-				VariableNode *variable;
-				ParameterNode *parameter;
-				IdentifierNode *bind;
+				VariableNode *variable = nullptr;
+				ParameterNode *parameter = nullptr;
+				IdentifierNode *bind = nullptr;
 			};
 			StringName name;
 			FunctionNode *source_function = nullptr;
@@ -1107,12 +1107,12 @@ public:
 		PropertyStyle property = PROP_NONE;
 		union {
 			FunctionNode *setter = nullptr;
-			IdentifierNode *setter_pointer;
+			IdentifierNode *setter_pointer = nullptr;
 		};
 		IdentifierNode *setter_parameter = nullptr;
 		union {
 			FunctionNode *getter = nullptr;
-			IdentifierNode *getter_pointer;
+			IdentifierNode *getter_pointer = nullptr;
 		};
 
 		bool exported = false;
diff --git a/platform/iphone/joypad_iphone.h b/platform/iphone/joypad_iphone.h
index 37e272a2c9..9d3c7d4589 100644
--- a/platform/iphone/joypad_iphone.h
+++ b/platform/iphone/joypad_iphone.h
@@ -40,7 +40,7 @@
 
 class JoypadIPhone {
 private:
-	JoypadIPhoneObserver *observer;
+	JoypadIPhoneObserver *observer = nullptr;
 
 public:
 	JoypadIPhone();
diff --git a/platform/osx/godot_content_view.h b/platform/osx/godot_content_view.h
index 7942d716dc..24ab7ac391 100644
--- a/platform/osx/godot_content_view.h
+++ b/platform/osx/godot_content_view.h
@@ -47,8 +47,8 @@
 
 @interface GodotContentView : RootView <NSTextInputClient> {
 	DisplayServer::WindowID window_id;
-	NSTrackingArea *tracking_area;
-	NSMutableAttributedString *marked_text;
+	NSTrackingArea *tracking_area = nullptr;
+	NSMutableAttributedString *marked_text = nullptr;
 	bool ime_input_event_in_progress;
 	bool mouse_down_control;
 	bool ignore_momentum_scroll;
diff --git a/servers/physics_2d/godot_body_pair_2d.h b/servers/physics_2d/godot_body_pair_2d.h
index 1c0f61be26..c9141dd220 100644
--- a/servers/physics_2d/godot_body_pair_2d.h
+++ b/servers/physics_2d/godot_body_pair_2d.h
@@ -40,8 +40,8 @@ class GodotBodyPair2D : public GodotConstraint2D {
 	};
 	union {
 		struct {
-			GodotBody2D *A;
-			GodotBody2D *B;
+			GodotBody2D *A = nullptr;
+			GodotBody2D *B = nullptr;
 		};
 
 		GodotBody2D *_arr[2] = { nullptr, nullptr };
diff --git a/servers/physics_2d/godot_joints_2d.h b/servers/physics_2d/godot_joints_2d.h
index acaaf0f629..a73e9333f4 100644
--- a/servers/physics_2d/godot_joints_2d.h
+++ b/servers/physics_2d/godot_joints_2d.h
@@ -76,8 +76,8 @@ public:
 class GodotPinJoint2D : public GodotJoint2D {
 	union {
 		struct {
-			GodotBody2D *A;
-			GodotBody2D *B;
+			GodotBody2D *A = nullptr;
+			GodotBody2D *B = nullptr;
 		};
 
 		GodotBody2D *_arr[2] = { nullptr, nullptr };
@@ -107,8 +107,8 @@ public:
 class GodotGrooveJoint2D : public GodotJoint2D {
 	union {
 		struct {
-			GodotBody2D *A;
-			GodotBody2D *B;
+			GodotBody2D *A = nullptr;
+			GodotBody2D *B = nullptr;
 		};
 
 		GodotBody2D *_arr[2] = { nullptr, nullptr };
@@ -141,8 +141,8 @@ public:
 class GodotDampedSpringJoint2D : public GodotJoint2D {
 	union {
 		struct {
-			GodotBody2D *A;
-			GodotBody2D *B;
+			GodotBody2D *A = nullptr;
+			GodotBody2D *B = nullptr;
 		};
 
 		GodotBody2D *_arr[2] = { nullptr, nullptr };
diff --git a/servers/physics_3d/godot_body_pair_3d.h b/servers/physics_3d/godot_body_pair_3d.h
index 7a7309f9d5..156db294d0 100644
--- a/servers/physics_3d/godot_body_pair_3d.h
+++ b/servers/physics_3d/godot_body_pair_3d.h
@@ -76,8 +76,8 @@ class GodotBodyPair3D : public GodotBodyContact3D {
 
 	union {
 		struct {
-			GodotBody3D *A;
-			GodotBody3D *B;
+			GodotBody3D *A = nullptr;
+			GodotBody3D *B = nullptr;
 		};
 
 		GodotBody3D *_arr[2] = { nullptr, nullptr };
diff --git a/servers/physics_3d/joints/godot_cone_twist_joint_3d.h b/servers/physics_3d/joints/godot_cone_twist_joint_3d.h
index fdcc2ceea3..6e111e7b6a 100644
--- a/servers/physics_3d/joints/godot_cone_twist_joint_3d.h
+++ b/servers/physics_3d/joints/godot_cone_twist_joint_3d.h
@@ -63,8 +63,8 @@ public:
 
 	union {
 		struct {
-			GodotBody3D *A;
-			GodotBody3D *B;
+			GodotBody3D *A = nullptr;
+			GodotBody3D *B = nullptr;
 		};
 
 		GodotBody3D *_arr[2] = { nullptr, nullptr };
diff --git a/servers/physics_3d/joints/godot_generic_6dof_joint_3d.h b/servers/physics_3d/joints/godot_generic_6dof_joint_3d.h
index bcf2d18647..902eebda55 100644
--- a/servers/physics_3d/joints/godot_generic_6dof_joint_3d.h
+++ b/servers/physics_3d/joints/godot_generic_6dof_joint_3d.h
@@ -143,8 +143,8 @@ class GodotGeneric6DOFJoint3D : public GodotJoint3D {
 protected:
 	union {
 		struct {
-			GodotBody3D *A;
-			GodotBody3D *B;
+			GodotBody3D *A = nullptr;
+			GodotBody3D *B = nullptr;
 		};
 
 		GodotBody3D *_arr[2] = { nullptr, nullptr };
diff --git a/servers/physics_3d/joints/godot_hinge_joint_3d.h b/servers/physics_3d/joints/godot_hinge_joint_3d.h
index b934540e8d..057cce13b5 100644
--- a/servers/physics_3d/joints/godot_hinge_joint_3d.h
+++ b/servers/physics_3d/joints/godot_hinge_joint_3d.h
@@ -56,8 +56,8 @@ subject to the following restrictions:
 class GodotHingeJoint3D : public GodotJoint3D {
 	union {
 		struct {
-			GodotBody3D *A;
-			GodotBody3D *B;
+			GodotBody3D *A = nullptr;
+			GodotBody3D *B = nullptr;
 		};
 
 		GodotBody3D *_arr[2] = {};
diff --git a/servers/physics_3d/joints/godot_pin_joint_3d.h b/servers/physics_3d/joints/godot_pin_joint_3d.h
index eeeaa650bd..db7f61dbe6 100644
--- a/servers/physics_3d/joints/godot_pin_joint_3d.h
+++ b/servers/physics_3d/joints/godot_pin_joint_3d.h
@@ -56,8 +56,8 @@ subject to the following restrictions:
 class GodotPinJoint3D : public GodotJoint3D {
 	union {
 		struct {
-			GodotBody3D *A;
-			GodotBody3D *B;
+			GodotBody3D *A = nullptr;
+			GodotBody3D *B = nullptr;
 		};
 
 		GodotBody3D *_arr[2] = {};
diff --git a/servers/physics_3d/joints/godot_slider_joint_3d.h b/servers/physics_3d/joints/godot_slider_joint_3d.h
index f596c9ff75..736dbee4f2 100644
--- a/servers/physics_3d/joints/godot_slider_joint_3d.h
+++ b/servers/physics_3d/joints/godot_slider_joint_3d.h
@@ -69,8 +69,8 @@ class GodotSliderJoint3D : public GodotJoint3D {
 protected:
 	union {
 		struct {
-			GodotBody3D *A;
-			GodotBody3D *B;
+			GodotBody3D *A = nullptr;
+			GodotBody3D *B = nullptr;
 		};
 
 		GodotBody3D *_arr[2] = { nullptr, nullptr };
```

bruvzg commented on platform/osx/godot_content_view.h:
```diff
@@ -47,8 +47,8 @@
 	DisplayServer::WindowID window_id;
-	NSTrackingArea *tracking_area;
-	NSMutableAttributedString *marked_text;
+	NSTrackingArea *tracking_area = nullptr;
+	NSMutableAttributedString *marked_text = nullptr;
```
Objective-C stuff can't be initialized in the definition. Both of these are already set in the constructor (`- (id)init`), so it's redundant anyway.

reduz commented:
Looks good in general, likely better to have a few more reviews, like @RandomShaper just to make sure I did not miss anything.

lawnjelly commented:
I did a little test just to see how well the compiler optimizes these out when they are unneeded, turns out it seems to be pretty good (in -O3 at least):

In godbolt
```
struct STest
{
    int myvar = 1;
};

int square(int num) {
    STest t;
    t.myvar = num;
    return t.myvar * t.myvar;
}

int test()
{
    STest t;
    return t.myvar * t.myvar;
}

int run()
{
    int res = square(5);
    return res;
}
```
Compiles to:
```
square(int):
        mov     eax, edi
        imul    eax, edi
        ret
test():
        mov     eax, 1
        ret
run():
        mov     eax, 25
        ret
```
This was actually the main thing I was worried about in performance sensitive areas, but it turns out if you set the internal value pretty soon after construction it is pretty good about eliminating the redundant set.

This may not work in all situations I guess (it probably depends how obvious it is to the optimizer) so it probably needs looking over carefully in the performance sensitive bits just in case.

akien-mga committed:
Zero initialize all pointer class and struct members

This prevents the pitfall of UB when checking if they have been
assigned something valid by comparing to nullptr.

akien-mga commented:
I tried some `clang-tidy` checks but I didn't have much luck to try to detect and fix cases where we initialize pointer members both in the header and in the constructor.

I tried:
- https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-prefer-member-initializer.htm - Doesn't seem to work properly (chokes on some of our constructs), and actually tries to move everything to list initialization instead of in-place.
- https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-pro-type-member-init.html - Finds some useful things to change, but it's actually doing other types of changes (like setting bools, etc.), so it's not really helping with removing unnecessary stuff.
- https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-default-member-init.html - We already applied this one a while ago, so the remaining cases are stuff it can't handle.
- https://clang.llvm.org/extra/clang-tidy/checks/readability-redundant-member-init.html - Despite the name it's only about calling parent class constructors explicitly when not needed. Doesn't hurt (will PR as a follow-up) but it's not related to this.

So I think the best might still be a manual review of all constructors (with half a dozen contributors working on it it shouldn't be too long) to finalize this modernization work started in 2020 :)

akien-mga commented:
I reviewed the core changes manually and cleaned up a bit, here's the diff (amended into this PR):
```diff
diff --git a/core/math/bvh.h b/core/math/bvh.h
index b0ea6cda80..f2c4edf5bc 100644
--- a/core/math/bvh.h
+++ b/core/math/bvh.h
@@ -763,9 +763,9 @@ private:
 		tree._extra[p_handle.id()].last_updated_tick = 0;
 	}
 
-	PairCallback pair_callback;
-	UnpairCallback unpair_callback;
-	CheckPairCallback check_pair_callback;
+	PairCallback pair_callback = nullptr;
+	UnpairCallback unpair_callback = nullptr;
+	CheckPairCallback check_pair_callback = nullptr;
 	void *pair_callback_userdata = nullptr;
 	void *unpair_callback_userdata = nullptr;
 	void *check_pair_callback_userdata = nullptr;
@@ -775,7 +775,7 @@ private:
 	// for collision pairing,
 	// maintain a list of all items moved etc on each frame / tick
 	LocalVector<BVHHandle, uint32_t, true> changed_items;
-	uint32_t _tick;
+	uint32_t _tick = 1; // Start from 1 so items with 0 indicate never updated.
 
 	class BVHLockedFunction {
 	public:
@@ -807,17 +807,10 @@ private:
 	Mutex _mutex;
 
 	// local toggle for turning on and off thread safety in project settings
-	bool _thread_safe;
+	bool _thread_safe = BVH_THREAD_SAFE;
 
 public:
 	BVH_Manager() {
-		_tick = 1; // start from 1 so items with 0 indicate never updated
-		pair_callback = nullptr;
-		unpair_callback = nullptr;
-		pair_callback_userdata = nullptr;
-		unpair_callback_userdata = nullptr;
-		_thread_safe = BVH_THREAD_SAFE;
-	}
 };
 
 #undef BVHTREE_CLASS
diff --git a/core/math/dynamic_bvh.h b/core/math/dynamic_bvh.h
index 32d1a7d434..74831089f3 100644
--- a/core/math/dynamic_bvh.h
+++ b/core/math/dynamic_bvh.h
@@ -215,10 +215,7 @@ private:
 			return axis.dot(volume.get_center() - org) <= 0;
 		}
 
-		Node() {
-			childs[0] = nullptr;
-			childs[1] = nullptr;
-		}
+		Node() {}
 	};
 
 	PagedAllocator<Node> node_allocator;
diff --git a/core/math/octree.h b/core/math/octree.h
index 56b475c0ea..65ab9e2292 100644
--- a/core/math/octree.h
+++ b/core/math/octree.h
@@ -156,18 +156,18 @@ private:
 	ElementMap element_map;
 	PairMap pair_map;
 
-	PairCallback pair_callback;
-	UnpairCallback unpair_callback;
+	PairCallback pair_callback = nullptr;
+	UnpairCallback unpair_callback = nullptr;
 	void *pair_callback_userdata = nullptr;
 	void *unpair_callback_userdata = nullptr;
 
-	OctreeElementID last_element_id;
-	uint64_t pass;
+	OctreeElementID last_element_id = 1;
+	uint64_t pass = 1;
 
-	real_t unit_size;
+	real_t unit_size = 1.0;
 	Octant *root = nullptr;
-	int octant_count;
-	int pair_count;
+	int octant_count = 0;
+	int pair_count = 0;
 
 	_FORCE_INLINE_ void _pair_check(PairData *p_pair) {
 		bool intersect = p_pair->A->aabb.intersects_inclusive(p_pair->B->aabb);
@@ -1265,18 +1265,7 @@ void Octree<T, use_pairs, AL>::set_unpair_callback(UnpairCallback p_callback, vo
 
 template <class T, bool use_pairs, class AL>
 Octree<T, use_pairs, AL>::Octree(real_t p_unit_size) {
-	last_element_id = 1;
-	pass = 1;
 	unit_size = p_unit_size;
-	root = nullptr;
-
-	octant_count = 0;
-	pair_count = 0;
-
-	pair_callback = nullptr;
-	unpair_callback = nullptr;
-	pair_callback_userdata = nullptr;
-	unpair_callback_userdata = nullptr;
 }
 
 #endif // OCTREE_H
diff --git a/core/object/message_queue.h b/core/object/message_queue.h
index a24ac3f306..2219cdb8f6 100644
--- a/core/object/message_queue.h
+++ b/core/object/message_queue.h
@@ -65,7 +65,7 @@ class MessageQueue {
 	uint8_t *buffer = nullptr;
 	uint32_t buffer_end = 0;
 	uint32_t buffer_max_used = 0;
-	uint32_t buffer_size;
+	uint32_t buffer_size = 0;
 
 	void _call_function(const Callable &p_callable, const Variant *p_args, int p_argcount, bool p_show_error);
 
diff --git a/core/os/os.h b/core/os/os.h
index 1b424ea011..b3c66ff18c 100644
--- a/core/os/os.h
+++ b/core/os/os.h
@@ -60,8 +60,6 @@ class OS {
 	bool _stdout_enabled = true;
 	bool _stderr_enabled = true;
 
-	char *last_error = nullptr;
-
 	CompositeLogger *_logger = nullptr;
 
 	bool restart_on_exit = false;
diff --git a/core/templates/thread_work_pool.h b/core/templates/thread_work_pool.h
index 363ea284fc..d364ac4fd8 100644
--- a/core/templates/thread_work_pool.h
+++ b/core/templates/thread_work_pool.h
@@ -49,7 +49,7 @@ class ThreadWorkPool {
 
 	template <class C, class M, class U>
 	struct Work : public BaseWork {
-		C *instance = nullptr;
+		C *instance;
 		M method;
 		U userdata;
 		virtual void work() {
```

fire-forge commented:
@akien-mga If the constructors need to be reviewed everywhere, the work could be split into chunks like in https://github.com/godotengine/godot/issues/58155. If you do that, I can take one of the chunks.
